module/shop/shop_voucher.py

```diff
@@ -0,0 +1,230 @@
+            row = 2
+        else:
+            logger.warning(f'Unexpected voucher icon match result: {[v.area for v in vouchers]}')
+            origin_y = 200
+            delta_y = 191
+            row = 2
+
+        # Make up a ButtonGrid
+        # Original grid is:
+        #shop_grid = ButtonGrid(
+        #     origin=(463, 200), delta=(156, 191), button_shape=(99, 99), grid_shape=(5, 2), name='SHOP_GRID')
+        shop_grid = ButtonGrid(
+            origin=(463, origin_y), delta=(156, delta_y), button_shape=(99, 99), grid_shape=(5, row), name='SHOP_GRID')
+        return shop_grid
+
+    @cached_property
+    def shop_voucher_items(self):
+        """
+        Returns:
+            ShopItemGrid:
+        """
+        shop_grid = self.shop_grid
+        shop_voucher_items = ShopItemGrid(
+            shop_grid,
+            templates={}, amount_area=(60, 74, 96, 95),
+            price_area=(52, 132, 132, 162))
+        shop_voucher_items.load_template_folder('./assets/shop/voucher')
+        shop_voucher_items.load_cost_template_folder('./assets/shop/cost')
+        shop_voucher_items.similarity = 0.85
+        shop_voucher_items.cost_similarity = 0.5
+        shop_voucher_items.price_ocr = PRICE_OCR
+        return shop_voucher_items
+
+    def shop_items(self):
+        """
+        Shared alias name for all shops,
+        so to use  @Config must define
+        a unique alias as cover
+
+        Returns:
+            ShopItemGrid:
+        """
+        return self.shop_voucher_items
+
+    def shop_currency(self):
+        """
+        Ocr shop voucher currency
+        Then return voucher count
+
+        Returns:
+            int: voucher amount
+        """
+        self._currency = OCR_SHOP_VOUCHER.ocr(self.device.image)
+        logger.info(f'Voucher: {self._currency}')
+        return self._currency
+
+    def shop_interval_clear(self):
+        """
+        Clear interval on select assets for
+        shop_buy_handle
+        """
+        super().shop_interval_clear()
+        self.interval_clear(SHOP_BUY_CONFIRM_SELECT)
+        self.interval_clear(SHOP_BUY_CONFIRM_AMOUNT)
+        self.interval_clear(SHOP_BUY_CONFIRM_VOUCHER)
+
+    def shop_buy_handle(self, item):
+        """
+        Handle shop_voucher buy interface if detected
+
+        Args:
+            item: Item to handle
+
+        Returns:
+            bool: whether interface was detected and handled
+        """
+        if self.appear(SHOP_BUY_CONFIRM_SELECT, offset=(20, 20), interval=3):
+            self.shop_buy_select_execute(item)
+            self.interval_reset(SHOP_BUY_CONFIRM_SELECT)
+            return True
+        if self.appear(SHOP_BUY_CONFIRM_AMOUNT, offset=(20, 20), interval=3):
+            self.shop_buy_amount_execute(item)
+            self.interval_reset(SHOP_BUY_CONFIRM_AMOUNT)
+            return True
+        if self.appear_then_click(SHOP_BUY_CONFIRM_VOUCHER, offset=(20, 20), interval=3):
+            return True
```

Comment:
Try to reuse `handle_popup_confirm()` as `SHOP_BUY_CONFIRM_VOUCHER`

Reply:
Did while debugging but just now realized I could temporarily update `_popup_offset` if that is acceptable?

I see in `retirement` does something similar `backup, self._popup_offset = self._popup_offset, (20, 50)`

Reply:
Add argument `offset` to `handle_popup_*` would be nice, see `handle_popup_single`. Modify dirty patches on `_popup_offset` too

Reply:
Done, updated func signatures and fixed dirty patches only `retirement` did this
Removed the unneeded asset